src/cmd/searchcontroller.cpp

```diff
@@ -1,66 +1,73 @@
+#include "timer.h"
 
-#include <cstdio>
+#include <algorithm>
+#include <future>
 #include <iostream>
+#include <thread>
+#include <utility>
 
-#include <boost/thread.hpp>
-#include <boost/timer.hpp>
-
-uint64_t readNext(FILE* file, char* buf, unsigned int blockSize) {
-  return std::fread(static_cast<void*>(buf), 1, blockSize, file);
+void print_cumulative_stats(double seconds, uint64_t offset) {
+  uint64_t units = offset >> 20;
+  const double bw = units / seconds;
+  units >>= 10;
+  std::cerr << units << " GB searched in "
+            << seconds << " seconds, "
+            << bw << " MB/s avg" << std::endl;
 }
 
 bool SearchController::searchFile(
-  std::shared_ptr<ContextHandle> searcher,
+  ContextHandle* searcher,
   HitCounterInfo* hinfo,
-  FILE* file,
+  Reader& reader,
   LG_HITCALLBACK_FN callback)
 {
-  boost::timer searchClock;
-  uint64_t blkSize = 0,
-         offset = 0;
+  Timer searchClock;
+  double thisTime, lastTime = searchClock.elapsed();
+
+  uint64_t blkSize,
+           offset = 0;
+
+  const char* buf;
+
+  std::tie(buf, blkSize) = reader.read(BlockSize).get();
+  while (blkSize) {
+    // start getting next block
+    std::future<std::pair<const char*, size_t>> fut = reader.read(BlockSize);
 
-  blkSize = readNext(file, Cur.get(), BlockSize);
-  if (!feof(file)) {
-    do {
-      // read the next block on a separate thread
-      boost::packaged_task<uint64_t> task(
-        std::bind(&readNext, file, Next.get(), BlockSize)
-      );
-      boost::unique_future<uint64_t> sizeFut = task.get_future();
-      boost::thread exec(std::move(task));
+    // search cur block
+    hinfo->setBuffer(buf, blkSize, offset);
 
-      // search cur block
-      lg_search(
-        searcher.get(),
-        Cur.get(),
-        Cur.get() + blkSize,
-        offset, hinfo, callback
-      );
+    lg_search(
+      searcher,
+      buf,
+      buf + blkSize,
+      offset, hinfo, callback
+    );
 
-      offset += blkSize;
-      if (offset % (1024 * 1024 * 1024) == 0) { // FIXME: should change this due to the block size being variable
-        const double lastTime = searchClock.elapsed();
-        uint64_t units = offset >> 20;
-        const double bw = units / lastTime;
-        units >>= 10;
-        std::cerr << units << " GB searched in " << lastTime
-                  << " seconds, " << bw << " MB/s avg" << std::endl;
-      }
-      blkSize = sizeFut.get(); // block on i/o thread completion
-      Cur.swap(Next);
-    } while (!feof(file)); // note file is shared btwn threads, but safely
+    offset += blkSize;
+
+    thisTime = searchClock.elapsed();
+    if (thisTime - lastTime > 30.0) {
+      print_cumulative_stats(searchClock.elapsed(), offset);
+      lastTime = thisTime;
+    }
+
+    std::tie(buf, blkSize) = fut.get(); // block on i/o thread completion
   }
 
   // assert: all data has been read, offset + blkSize == file size,
   // cur is last block
+  hinfo->setBuffer(buf, blkSize, offset);
+
   lg_search(
```

Comment:
This function call easily fits in 80 columns:
```
  lg_search(searcher, buf, buf + blkSize, offset, hinfo, callback);
```

Reply:
It once did not. Fixed.